great_expectations/agent/actions/list_table_names.py

```diff
@@ -0,0 +1,37 @@
+from typing import TYPE_CHECKING, List
+
+from great_expectations.agent.actions.agent_action import (
+    ActionResult,
+    AgentAction,
+)
+from great_expectations.agent.models import (
+    ListTableNamesEvent,
+)
+from great_expectations.compatibility.sqlalchemy import inspect
+from great_expectations.datasource.fluent import Datasource as FluentDatasource
+from great_expectations.datasource.fluent import SQLDatasource
+
+if TYPE_CHECKING:
+    from great_expectations.compatibility.sqlalchemy.engine import Inspector
+
+
+class ListTableNamesAction(AgentAction[ListTableNamesEvent]):
+    def run(self, event: ListTableNamesEvent, id: str) -> ActionResult:
+        datasource_name: str = event.datasource_name
+        datasource: FluentDatasource = self._context.get_datasource(
+            datasource_name=datasource_name
+        )
+        if not isinstance(datasource, SQLDatasource):
+            raise TypeError(
+                f"This operation requires a SQL Datasource but got {type(datasource).__name__}."
+            )
+
+        inspector: Inspector = inspect(datasource.get_engine())
+        table_names: List[str] = inspector.get_table_names()
```

Comment:
I'm interested in exploring how we could do this within the current agent pattern. Here's one possible solution:

- This logic is moved into the `SqlDatasource` FDS base class.
- `CloudDataContext` gets a new method `introspect_datasource`, which for FDS SQL datasources calls the method defined above, and otherwise raises an error that introspecting this datasource isn't supported (yet).
- After the data context receives the list of tables, it posts back to a new mercury endpoint which persists the results of the introspection, possibly on the datasource table itself.

It's very possible that our initial design is flawed, and that we'll need to implement something along the lines of what you proposed in this PR, but I think it would be useful to try an implementation within the initial spec first, to understand what its missing.

Reply:
I think it's useful to think about some of the agent jobs we plan to add in the near future.
Some of these almost certainly do not make sense as features/methods in the core OSS library.

For example, we are planning a `Get ENV VARS` agent job, that would read all the ENVIRONMENT VARIABLES it has access to and returns that list.

In the past, we have been very lax about adding to the surface area of the GX public API without much consideration. I imagine there's going to be a fair number of Agent features that we really don't want or need as part of the GX core API.